Lifi bridge integration to swapr

wixzi opened the pull request:
## Fixes: #1735 

# Description

Lifi bridge

![Screenshot 2023-03-11 at 16 21 35](https://user-images.githubusercontent.com/102727631/224492688-d4175e32-c8d6-4a86-85c8-e8dbb6d8d69a.png)
![Screenshot 2023-03-11 at 16 21 51](https://user-images.githubusercontent.com/102727631/224492690-b8b2da48-7b55-4fa1-8358-b22069f7425f.png)


# How to test the changes

Go to bridge page and see Lifi bridge is one of the options.

# Are there any risks or changes of failure to be considered

Low risks. Lifi is a stable bridge that utilze mulitiple routes to finish bridging. One issue is you cannot request history of bridges for an account done through Lifi

berteotti commented on src/services/EcoBridge/Lifi/Lifi.selectors.ts:
```diff
@@ -0,0 +1,152 @@
+const createSelectPendingTransactions = (selectOwnedTxs: ReturnType<typeof createSelectOwnedTransactions>) =>
+  createSelector(selectOwnedTxs, ownedTxs => {
+    const pendingTxs = ownedTxs.filter(
+      ({ statusResponse }) => statusResponse.status !== 'FAILED' && !statusResponse.receiving?.txHash
+    )
+
+    return pendingTxs
+  })
+
+const createSelectFailedTransactions = (selectOwnedTxs: ReturnType<typeof createSelectOwnedTransactions>) =>
+  createSelector(selectOwnedTxs, ownedTxs => {
+    const failedTxs = ownedTxs.filter(
+      ({ statusResponse }) => statusResponse.status === 'FAILED' && !statusResponse.receiving?.txHash
```
Use `LifiTxStatus` here instead of `'FAILED'`

berteotti commented on src/services/EcoBridge/Lifi/LifiBridge.ts:
```diff
@@ -0,0 +1,407 @@
+    const promises = lifiExecutedRoutes.map(async tx => {
+      try {
+        const { statusRequest, statusResponse } = tx
+        if (statusResponse.status !== 'DONE' && statusResponse.status !== 'FAILED') {
+          const result = await LifiApi.getStatus(statusRequest, { signal: this.#renewAbortController('status') })
+          if (result.status === 'DONE') {
+            this.store.dispatch(this.#actions().updateTx({ ...result, timeResolved: Date.now() }))
+          } else {
+            this.store.dispatch(this.#actions().updateTx(result))
+          }
+
+          if (result.status === 'FAILED' || result.status === 'INVALID' || result.status === 'NOT_FOUND') {
```
Shouldn't we use types here?

berteotti reviewed (commented):
Left some comments, code looks good and just need a few improvements

MilanVojnovic95 commented:
As far as I have noticed by testing on different networks and with different tokens, it appears everywhere as an option, however, bridging is not possible.

@wixzi Maybe there will be a special PR where we can test the bridging itself?

## Tested:
#### - USDT from Polygon to Arbitrum One ❌ 

![image](https://user-images.githubusercontent.com/57331244/224992295-126bd74e-87ea-48c2-be64-2ee0ec600dd9.png)

#### - USDT from Arbitrum One to Polygon ❌ 

![image](https://user-images.githubusercontent.com/57331244/224993057-1471ed4f-4429-4454-91c7-7d51022bff2c.png)

#### USDC from Optimism to Gnosis Chain ❌ 

![image](https://user-images.githubusercontent.com/57331244/224993665-db644bc5-5485-493c-952c-5e68e04443e6.png)

wixzi commented on src/services/EcoBridge/Lifi/Lifi.selectors.ts:
```diff
@@ -0,0 +1,152 @@
+        toValue,
+        txHash: statusResponse.receiving?.txHash!,
+        pendingReason,
+        // TODO: need to fix this
```
I gave `Date.now()` when DONE but that is should work for now since we are not fetching history. I will add comment to fetch the correct time from onChian

wixzi commented:
@MilanVojnovic95 I can test again and see. I was able to do bridging from polygon to optimism.

wixzi commented on src/services/EcoBridge/Lifi/LifiBridge.ts:
```diff
@@ -0,0 +1,445 @@
+      // const signTransaction = await signer.signTransaction(route.transactionRequest)
+
+      const transaction = await signer.sendTransaction(route.transactionRequest)
+      const transactionReceipt = await transaction.wait()
```
@adamazad Need some help here

wixzi replied:
@berteotti The transaction is failing here. Dont know why.

wixzi commented:
Bridging initiated
![Screenshot 2023-04-05 at 16 17 53](https://user-images.githubusercontent.com/102727631/230109976-1b605b7f-7931-48bc-9751-a0a74f50840a.png)

Bridging pending
![Screenshot 2023-04-05 at 16 18 03](https://user-images.githubusercontent.com/102727631/230109984-00ce4108-2718-419a-a5d1-ffcf373d7a5c.png)

Pending in bridging history
![Screenshot 2023-04-05 at 16 18 12](https://user-images.githubusercontent.com/102727631/230109987-32073d49-3725-42fc-acc4-0c952d3afcab.png)
Pending in account status also

![Screenshot 2023-04-05 at 16 18 19](https://user-images.githubusercontent.com/102727631/230109992-9014ac96-fb79-4755-9c73-acc057e01deb.png)

Polling status of the pending transaction
![Screenshot 2023-04-05 at 16 18 39](https://user-images.githubusercontent.com/102727631/230109998-660d9a4a-51ff-4b6e-8ad1-afe9dff49a22.png)

Confirmed transaction and updated result
![Screenshot 2023-04-05 at 16 20 09](https://user-images.githubusercontent.com/102727631/230110002-49a174e5-95ec-4edb-a53e-62cd40c1b4ca.png)
Confirmed transaction and updated result
![Screenshot 2023-04-05 at 16 20 18](https://user-images.githubusercontent.com/102727631/230110008-5d847bd9-8e49-4632-968c-49fbd04b8c85.png)
Confirmed transaction and updated result
![Screenshot 2023-04-05 at 16 20 28](https://user-images.githubusercontent.com/102727631/230110012-587d1b79-a364-44e1-8e87-154ea4d4f638.png)

berteotti commented on src/services/EcoBridge/Lifi/Lifi.reducer.ts:
```diff
@@ -0,0 +1,100 @@
+    name: bridgeId,
+    initialState,
+    reducers: {
+      // TODO: cleanup redux and fix txn history
```
I think this is done

berteotti commented on src/services/EcoBridge/Lifi/Lifi.reducer.ts:
```diff
@@ -0,0 +1,101 @@
+
+        state.transactions.forEach((txn, index) => {
+          if (txn.statusResponse.sending.txHash === statusResponse.sending.txHash) {
+            state.transactions[index] = { ...txn, statusResponse: statusResponse, timeResolved }
```
No need for declare `statusResponse` twice but either it's ok
```suggestion
            state.transactions[index] = { ...txn, statusResponse, timeResolved }
```

berteotti commented on src/services/EcoBridge/Lifi/LifiBridge.ts:
```diff
@@ -0,0 +1,448 @@
+      tokens: [],
+    }
+
+    // if (isBridgeSwapActive) {
```
Is this comment necessary? Is the bridge capable of bridge swaping?

berteotti commented on src/services/EcoBridge/Lifi/LifiBridge.ts:
```diff
@@ -0,0 +1,448 @@
+
+      if (!this._account || !checkRoute || !this._activeProvider || !route.transactionRequest) return
+      const signer = this._activeProvider.getSigner()
+      // const checkTransaction = await signer.populateTransaction(route.transactionRequest)
+      // const signTransaction = await signer.signTransaction(route.transactionRequest)
```
This comments can be deleted

berteotti reviewed (approved):
PR looks good to me, just small typo fixes and code cleaning needed. Good job Velu!

wixzi commented:
@berteotti Thank you 🙏🏼
